swift/ql/lib/codeql/swift/security/XXE.qll

```diff
@@ -163,3 +164,40 @@ private class AexmlOptions extends Expr {
     this.getType() = any(LValueType t | t.getObjectType() instanceof AexmlOptionsType)
   }
 }
+
+/** The XML argument of a `libxml2` parsing call vulnerable to XXE. */
+private class Libxml2XxeSink extends XxeSink {
+  Libxml2XxeSink() {
+    exists(Libxml2ParseCall c, Libxml2BadOption opt |
+      this.asExpr() = c.getXml() and
+      lib2xmlOptionLocalTaintStep*(DataFlow::exprNode(opt.getAnAccess()),
+        DataFlow::exprNode(c.getOptions()))
+    )
+  }
+}
+
+/**
+ * Holds if taint can flow from `source` to `sink` in one local step,
+ * including bitwise operations, accesses to `.rawValue`, and casts to `Int32`.
+ */
+private predicate lib2xmlOptionLocalTaintStep(DataFlow::Node source, DataFlow::Node sink) {
```

Comment:
The bitwise operation and `.rawValue` ones are probably of general interest as taint steps. The `Int32` constructor technically applies globally too, although it may seem a little more niche.

Reply:
Addressed in https://github.com/github/codeql/pull/14502 .  Actually, we've chipped away at this in a few steps, but that PR gets rid of the last special case here.